### PR TITLE
Let the user choose prefixes output type 

### PR DESCRIPTION
**Pull request by tigrulya-exe**
- Now excluded prefixes collector delegates writing the prefixes to provided writer.
- Added configmap and file output prefixes writers. The last one is used by default.
- Added tests for file output case.

### CODE REVIEW

File: internal/prefixcollector/collector.go [hunk 1]
```diff
@@ -21,115 +21,62 @@ import (
...
+type PrefixSource interface {
 	Prefixes() []string
 }
 
-// ExcludePrefixCollector is service, collecting excluded prefixes from list of ExcludePrefixSource
+// PrefixesWriter is excluded prefixes writer
+type PrefixesWriter interface {
+	Write(context.Context, []string)
+	WatchExcludedPrefixes(context.Context, *utils.SynchronizedPrefixesContainer)
+}
```
Comment (denis-tingajkin) [inline internal/prefixcollector/collector.go:38]: Can we just pass a function into the collector that means how to write sources?
```suggestion
```

```go
type WritePrefixesFunc func(Config, []strings)
```

**Review by denis-tingajkin — CHANGES_REQUESTED** [review]

File: internal/prefixcollector/collector.go (hunk 1 above)
Comment (tigrulya-exe) [reply internal/prefixcollector/collector.go]: Replaced writers with WritePrefixesFunc

File: internal/prefixcollector/collector.go [hunk 2]
```diff
@@ -23,113 +23,66 @@ import (
...
-func NewExcludePrefixCollector(notifyChan <-chan struct{}, configMapName, configMapNamespace string,
-	sources ...ExcludePrefixSource) *ExcludePrefixCollector {
-	return &ExcludePrefixCollector{
-		nsmConfigMapName:      configMapName,
-		nsmConfigMapNamespace: configMapNamespace,
-		notifyChan:            notifyChan,
-		sources:               sources,
-		previousPrefixes:      utils.NewSynchronizedPrefixesContainer(),
+func NewExcludePrefixCollector(notifyChan <-chan struct{}, writeFunc WritePrefixesFunc,
+	watchFunc WatchPrefixesFunc, sources ...PrefixSource) *ExcludePrefixCollector {
```
Comment (denis-tingajkin) [inline internal/prefixcollector/collector.go:54]: I still think that `options` here can make code some clear.

For example:

```suggestion
func NewExcludePrefixCollector(options...Option) *ExcludePrefixCollector {
```

Then main.go could look something like:

```go
func main() {
    ...
    var options []Option
    if config.OutputFilePath != "" {
        options = append(options, WithOutputFile(config.OutputFilePath))
    }
    ...
    go NewExcludePrefixCollector(options...).Serve(ctx)
    <-ctx.Done()
}
```

File: main.go [hunk 3]
```diff
@@ -121,3 +134,21 @@ func validatedPrefixes(prefixes []string) ([]string, error) {
...
 }
+
+func getPrefixesWriteFunc(config *Config, currentNamespace string) (prefixcollector.WritePrefixesFunc, error) {
+	switch config.PrefixesOutputType {
+	case fileOutputType:
+		return prefixwriter.NewFileWriter(config.OutputFilePath), nil
+	case configMapOutputType:
+		return prefixwriter.NewConfigMapWriter(config.NSMConfigMapName, currentNamespace), nil
+	default:
+		return nil, errors.New("Wrong prefixes output")
```
Comment (denis-tingajkin) [inline main.go:145]: I would also suggest adding method Validate for the Config

Comment (tigrulya-exe) [reply main.go]: Added validate method

File: internal/prefixcollector/collector.go (hunk 2 above)
Comment (tigrulya-exe) [reply internal/prefixcollector/collector.go]: Added options for collector

File: internal/prefixcollector/collector.go [hunk 4]
```diff
@@ -23,113 +23,103 @@ import (
...
+	}
+}
+
+// NewExcludePrefixCollector creates ExcludedPrefixCollector
+func NewExcludePrefixCollector(options ...Option) *ExcludedPrefixCollector {
+	collector := &ExcludedPrefixCollector{
+		notifyChan:       make(chan struct{}, 1),
+		previousPrefixes: utils.NewSynchronizedPrefixesContainer(),
+		writeFunc:        fileWriter(defaultPrefixesFilePath),
+		watchFunc:        nil,
```
Comment (denis-tingajkin) [inline internal/prefixcollector/collector.go:93]:
```suggestion
```

File: main.go [hunk 5]
```diff
@@ -85,39 +76,39 @@ func main() {
...
 	<-ctx.Done()
 }
 
-// validatedPrefixes returns list of validated via CIDR notation parsing prefixes
-func validatedPrefixes(prefixes []string) ([]string, error) {
-	var validatedPrefixes []string
-	for _, prefix := range prefixes {
-		_, _, err := net.ParseCIDR(prefix)
+func getPrefixesOutputOption(config *prefixcollector.Config) (prefixcollector.Option, error) {
+	if config.PrefixesOutputType == prefixcollector.ConfigMapOutputType {
```
Comment (denis-tingajkin) [inline main.go:104]: Can we move this condition into main func?
```suggestion
```

**Review by denis-tingajkin — COMMENTED** [review]
Added a few minor comments. In general looks good.

File: main.go (hunk 5 above)
Comment (tigrulya-exe) [reply main.go]: Moved

File: internal/prefixcollector/collector.go (hunk 4 above)
Comment (tigrulya-exe) [reply internal/prefixcollector/collector.go]: Removed

**Review by denis-tingajkin — APPROVED** [review]

**Pull request merged**